### ISSUE TITLE
ENH: Expose `bit_count` ufunc equivalent to the scalar methods

### ISSUE THREAD

**Issue by hexchain** [numpy/numpy#16325]
Modern CPUs have a builtin [popcount](https://en.wikichip.org/wiki/population_count) instruction that counts the number of 1 bit in an unsigned integer. It would be great if numpy has a routine that exposes this operation.

Both GCC and MSVC have builtin functions for this instruction.

Possible usage:
```python
>>> np.popcount(4)
1
>>> np.popcount(np.arange(5, dtype=np.uint))
array([0, 1, 1, 2, 1])
```

**Comment by seberg** [comment]
I would probably argue for calling it `countbits` or so to stay close with `unpackbits` and `packbits` (although that means `countnonzero` is similar).
The only discussion may be around whether or not it adds new API, I think a PR adding a ufunc for this has very good chances to be accepted (but should go through the mailing list since its new API).

**Comment by eric-wieser** [comment]
Of course, while we're playing this game we may also want to expose some other intrinsics from https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html:

* `int __builtin_ffs (int x)`
  Returns one plus the index of the least significant 1-bit of `x`, or if `x` is zero, returns zero.

* `int __builtin_clz (unsigned int x)`
  Returns the number of leading 0-bits in `x`, starting at the most significant bit position. If `x` is 0, the result is undefined.

* `int __builtin_ctz (unsigned int x)`
  Returns the number of trailing 0-bits in `x`, starting at the least significant bit position. If `x` is 0, the result is undefined.

* `int __builtin_clrsb (int x)`
  Returns the number of leading redundant sign bits in `x`, i.e. the number of bits following the most significant bit that are identical to it. There are no special cases for 0 or other values.

* `int __builtin_popcount (unsigned int x)`
  Returns the number of 1-bits in `x`.

* `int __builtin_parity (unsigned int x)`
  Returns the parity of `x`, i.e. the number of 1-bits in `x` modulo 2.

**Comment by hexchain** [comment]
Yes, `countbits` sounds like a better name.

I would like to put up a PR but I'm still trying to get familiar with numpy codebase. Besides, there are some other questions:

- `__builtin_popcount(x)` only works with unsigned integer. What should `np.countbits` return if some signed integers are given? Should it treat them as unsigned implicitly (like in C) or should it error out?
- Should we implement our own `popcount` using a loop in case such builtins are not available in the compiler? (Also, should we do the same for every other compiler builtins in the future?)

**Comment by seberg** [comment]
Yeah, I guess those other intrinsics are the exact reason why we likely would not want to include it unless `countbits` is somehow much more widely applicable than the others...

At least unless we add a new namespace somewhere. Or maybe this is a good reason for creating a package containing additional ufuncs...

**Comment by eric-wieser** [comment]
> * `__builtin_popcount(x)` only works with unsigned integer. What should `np.countbits` return if some signed integers are given? Should it treat them as unsigned implicitly (like in C) or should it error out?

If it's a ufunc, you'd just declare that the ufunc only accepts the signed integer types

> * Should we implement our own `popcount` using a loop in case such builtins are not available in the compiler? (Also, should we do the same for every other compiler builtins in the future?)

If we're going to add the function, it needs to behave on all platforms. Perhaps we should detect whether the builtin is supported at compile-time.

Also, you should add a loop that works on regular python integers which are larger than `long long`.

**Comment by eric-wieser** [comment]
> Yeah, I guess those other intrinsics are the exact reason why we likely would not want to include it unless `countbits` is somehow much more widely applicable than the others...

For what it's worth, I found myself needing `countbits` recently in pygae/clifford, but since I was using numba anyway it didn't really matter.

> At least unless we add a new namespace somewhere. Or maybe this is a good reason for creating a package containing additional ufuncs...

I was thinking this too. It would be a nice excuse to force ourselves to make a new namespace.
An argument for not having an additional package is it might be harder to hook into the compiler-config testing that numpy does in setup.py.

**Comment by hexchain** [comment]
> Yeah, I guess those other intrinsics are the exact reason why we likely would not want to include it unless `countbits` is somehow much more widely applicable than the others...

I happen to be in need of such a thing when working on a project. I've done some search and found other people trying to achieve the same thing, and there is even [a CPython enhancement request](https://bugs.python.org/issue29882). So I guess it would be worth adding, and because we have C-level loops, it can work on an array of numbers more efficiently than the [proposed](https://github.com/python/cpython/pull/771) CPython implementation.

As a comparison of popularity, I did some Google search to see if there are people interested in the functions of these `__builtin` routines. It seems that judging by the relevance of top results and the total number of search results:
- `popcount` ("python count one bits"): 4070000 results
- `ctz` ("python count trailing zero bits"): 384000 results
- `clz` ("python count leading zero bits"): 2240000 results
- `parity` ("python integer parity"): 1040000 results
- (I couldn't think of a clear description of the other two, and searching by their names does not lead to anything relevant)

The result might be off due to different search words, but I believe `popcount` should still be the most popular one.

**Comment by BartMassey** [comment]
I'm about to write some code in another language because I don't have any convenient way to access a popcount primitive from Python and I need the performance, for whatever that's worth. Strongly support this.

**Comment by eric-wieser** [comment]
`numba` doesn't make it too hard to use popcnt, if you're looking for an excuse not to leave python.

**Comment by j-towns** [comment]
I would also find NumPy popcount/countbits useful.

In case it's helpful for anyone: as well as Numba it's also available in JAX, as [`jax.lax.population_count`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.population_count.html#jax.lax.population_count), and here's an implementation in terms of other NumPy functions: https://github.com/google/jax/blob/6c8fc1b031275c85b02cb819c6caa5afa002fa1d/jax/lax_reference.py#L121-L150.

**Comment by mborgerding** [comment]
> As a comparison of popularity, I did some Google search to see if there are people interested in the functions of these `__builtin` routines. It seems that judging by the relevance of top results and the total number of search results:
> 
> * `popcount` ("python count one bits"): 4070000 results
> * `parity` ("python integer parity"): 1040000 results

Note parity is trivial to implement if you have an efficient popcount. `(parity=popcount&1)`

**Comment by BartMassey** [comment]
> here's an implementation in terms of other NumPy functions: https://github.com/google/jax/blob/6c8fc1b031275c85b02cb819c6caa5afa002fa1d/jax/lax_reference.py#L121-L150.

This is an implementation of divide-and-conquer popcount, which is going to be substantially slower than something that references native popcount instructions. http://github.com/BartMassey/popcount shows 4× slowdown on native code, but the Python implementation will be way slower than that I suspect.

**Comment by j-towns** [comment]
Haha yes I wasn’t claiming it was efficient 🙂. Perhaps I should have added an explicit warning.

**Comment by mborgerding** [comment]
Here's a numpy-only approach I coded up this morning (to fill the need that lead me here)

I have not profiled it, but it made my simulation acceptably fast.

```python
# create a lookup table for set bits per byte
lookuptable = np.array([bin(i).count('1') for i in range(256)]).astype(np.int32)
def count_set_bits(v):
    'returns the count of set bits in each element of v'
    assert v.dtype in (np.uint8, np.uint16, np.uint32, np.uint64), 'must be an unsigned int dtype'
    return lookuptable[np.reshape(v.view(np.uint8), v.shape + (-1, ))].sum(axis=-1)
```

**Comment by BartMassey** [comment]
Table-driven is probably the best choice for this situation, and this is a nice implementation. Thanks!

That said, still want a `popcount` primitive. :-)

**Comment by ganesh-k13** [comment]
Hi All, is this open for a PR? Would like to try adding a new ufunc, having touched upon it briefly in my previous PRs.

**Comment by mborgerding** [comment]
> Hi All, is this open for a PR? Would like to try adding a new ufunc, having touched upon it briefly in my previous PRs.

If my `count_set_bits` code above is helpful, please use it.  It is easier for me to find the time to post a code snippet or a comment than to do a PR.

**Comment by mattip** [comment]
Before we add this to NumPy, we need to have an open discussion, preferably on the mailing list, about expanding the NumPy API. In general, I think the approach should be to stick to the [array API spec](https://data-apis.org/array-api/latest/API_specification/index.html) or at least have a good reason for deviating from it. I see JAX is mentioned here, does anyone know if there was discussion of this function (or other compiler intrinsics) in the array API spec design?

**Comment by ganesh-k13** [comment]
Thanks for the info Matti, was not aware of the array API spec. Yeah, it makes sense to reduce duplicate efforts.
Here are a few points I noticed that might make help the case to include popcount here:
1. We can provide a native C implementation and use built-in popcounts where we can. The main advantage over the JAX implemenation being the one in current PR uses 12 instructions [[1](https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel)], lesser than that of JAX. That being said, we can improve the JAX implementation or we can move this to SciPy.
2. With SIMD support, we can get massive improvements in large array cases with builtin SSE and AVX popcounts.
So I guess, we can take a call on where this should go, will be happy to shift it to SciPy, etc or improve some existing implementation.

Should I start a mail thread? Not sure what all to include in the thread :). Any pointers would help.

[1]: https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel

**Comment by eric-wieser** [comment]
Some other things to consider here:

* There is already naming precedent for `bit_count` instead of `count_bits` - the builtin python method `int.bit_count`
* CPython has a builtin implementation of `popcount`, `_Py_popcount32` - but it obviously isn't available for 64-bit integers.

**Comment by rgommers** [comment]
> In general, I think the approach should be to stick to the [array API spec](https://data-apis.org/array-api/latest/API_specification/index.html) or at least have a good reason for deviating from it.

It's not in there, and nor are `packbits`/`unpackbits` (too niche, and not widely enough implemented in other array/tensor libraries). The one thing I think may be relevant is that the lfunctions that are in there that deal with bitwise operations chose a uniform `bitwise_*` naming scheme. `numpy` has `bitwise_and/or/xor/not`, but then also has `left_shift`, `right_shift`, `invert` - which is all a bit messy, and something like `invert` doesn't make too much sense as a name. 

We have to have support for functions that are Python operators, but in an ideal world I think all bit-twiddly functions would be in a separate namespace or package.

**Comment by eric-wieser** [comment]
I think that the fact python itself has `int.count_bits()` is a good argument for numpy (if not the array API) supporting this.

**Comment by rgommers** [comment]
@eric-wieser just to clarify, you do want a new ufunc right, you are not suggesting adding a `bit_count` method?

It seems like there's enough interest, so it seems to make sense to me to add it - the question is just under what name.

> if not also the array API

Also for clarificaiton: the array API standard is about ~125 functions, largely a synthesis of what's already in most/all of NumPy/TensorFlow/PyTorch/JAX/Dask/CuPy/MXNet. So functions like these, which are new and have a fairly small audience compared to the total user population of those libraries, will not be considered for inclusion any time soon I think. What it can do though is help align around naming/signature/semantics, so if it gets added in the future it will be consistent with what's already there (and it helps if we make design choices visible to TensorFlow, PyTorch et al. so they don't add the same thing with a different name).

**Comment by eric-wieser** [comment]
Frankly I think we should add both; a new `bit_count` ufunc, and a `np.integer.bit_count()` wrapper just to make numpy ints work slightly better where `int` is expected

**Comment by rgommers** [comment]
I like the name `bit_count`. Certainly much more descriptive than `popcount`.

**Comment by seberg** [comment]
`bit_count` seems great (probably suits a bit better to methods, but its nice that it starts with `bit`, makes it easier to find).  Pythons version uses `abs(int)` for negative ones, we may have to take a bit of care if we support signed ints.

I would also be happy to have a `np.bitwise` or `np.lib.bitwise` or similar namespace that pools all bit "fiddling" functions (which may mean duplicating).   But that probably needs some dedicated discussion.

---

@rgommers I am struggling with the data-apis role here.  It is not that there are too many steps involved.  But it would be awesome if data-api's could make these type of decisions easier (as in a clear, simple steps to "involve" it).  But for me, it adds a bit of fuzziness since we all like to think that data-apis *should* be involved in some form, but it is not really clear how.  (It helps a lot with existing functions that seem settled of course, but that doesn't apply to most API changes.)

There is just nothing that clearly helps much for API additions?  Probably API comparison lists do (i.e. to find that JAX already has (`population_count`)[https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.population_count.html#jax.lax.population_count]).  But even there I do not actually know where I should be looking.  (Which is fine, I don't expect!)

There is also no mechanism (besides the core people) to exchange information about *new* API.  Either the core data-api people ensure that the information is forwarded to whoever should get it (which means that we do our part by tagging and mailing list announcements and maybe pining you or Stephan), or we develop something else:
* A new data-api "discuss", so it is easy for anyone to say: `bit_count` sounds good. It may be interesting for similar libs, so lets also post it at XY.  (Or a python.org hosted mailing list)
* A github scrapper(?) for tagged issues (maybe even a new tag), so that its easy to follow "important" API decisions across projects.

It is great to add a "cross project" layer of discussion to some API decisions (it may also be good to include downstream a bit more).  But to me it is completely unclear how that layer currently works/looks like if I assume that data-api's is that layer.

**Comment by rgommers** [comment]
>  But it would be awesome if data-api's could make these type of decisions easier (as in a clear, simple steps to "involve" it). But for me, it adds a bit of fuzziness since we all like to think that data-apis _should_ be involved in some form, but it is not really clear how. (It helps a lot with existing functions that seem settled of course, but that doesn't apply to most API changes.)

Yes I agree, very good points @seberg. There's both a need to define better what the scope is for these kind of future-looking decisions on new API, and a simple "here's how to get involved, here's steps to take to align with other libraries on a new function".

> There is just nothing that clearly helps much for API additions? Probably API comparison lists do ....
> But even there I do not actually know where I should be looking. (Which is fine, I don't expect!)

There's actually a whole repo for that, with hopefully good enough instructions to show all the data in current API comparisons: https://github.com/data-apis/array-api-comparison.


> A new data-api "discuss", so it is easy for anyone to say: `bit_count` sounds good. It may be interesting for similar libs, so lets also post it at XY.  (Or a python.org hosted mailing list)

Yes, good idea. We signed up early for the GitHub Discussions beta, but it seems no one looks there - I've never found a useful Discussions tab on a repo (unfortunately). So it's probably either a mailing list or a Discourse. Given it's 2021, probably a Discourse is best?

**Comment by seberg** [comment]
@rgommers we can also use https://discuss.scientific-python.org/ as Stéfan pointed out.  That would not be as focused on array-libs, but that may be just as well/better since there are abound to be many interested e.g. in NumPy API and sklearn API...

The big advantage of all of this is that all the tagging in the world on github doesn't really help much: It is hard to get notifications for specific tags.  But discuss/discourse has nested topics, so it is easy to do `APIs/numpy`, `APIs/scipy`, and you can "watch" all projects or just the projects that interest you in one place.

It may not be as "Data-apis" centric, but I think it would fit well enough:  Each projects representatives can just add a tag, and those interested can watch the tag just as you can watch a topic.

**Comment by BartMassey** [comment]
The discussion of meta-process for this project has been fascinating. That said, we still would like a `popcount()` function. What's the minimum-friction way to get one in as quickly as reasonably possible?

**Comment by mattip** [comment]
It will be available when #19355 is merged.

**Comment by BartMassey** [comment]
Thanks much!

**Comment by seberg** [comment]
Reopening, since we did not add the ufunc version yet, it is a valid request to add that.

**Comment by vadimkantorov** [comment]
@rgommers Seems popcnt is useful for some binary neural nets: https://github.com/pytorch/pytorch/issues/32867

**Comment by vscodegithubby** [comment]
All the c++ intrinsics should be added:
countl_zero
countl_one
countr_zero
countr_one
has_single_bit
bit_count
